### PR TITLE
Add partitioner for n-level graph partitioning

### PR DESCRIPTION
**Pull request by N-Maas**
Additional refactorings along the way:
- extracted some functionality from `dynamic_hypergraph` to `contraction_tree`
- made `copy()` `const` for all graph data structures

### CODE REVIEW

File: mt-kahypar/datastructures/dynamic_adjacency_array.cpp [hunk 1]
```diff
@@ -0,0 +1,672 @@
...
+  ASSERT(verifyIteratorPointers(v), "Iterator pointers of vertex" << v << "are corrupted");
+}
+
+void DynamicAdjacencyArray::streamWeight(StreamingVector<RemovedEdgesOrWeight>& tmp_removed_edges,
+                                         const ParallelEdgeInformation& e, HyperedgeWeight w) {
+  tmp_removed_edges.stream(RemovedEdgesOrWeight::asWeight(e.header_id, edge(e.edge_id).weight, e.target));
+  edge(e.edge_id).weight += w;
+}
+
+parallel::scalable_vector<DynamicAdjacencyArray::RemovedEdgesOrWeight> DynamicAdjacencyArray::removeSinglePinAndParallelEdges() {
```
Comment (kittobi1992) [inline mt-kahypar/datastructures/dynamic_adjacency_array.cpp:312]: as we have discussed ...

Comment (kittobi1992) [inline mt-kahypar/datastructures/dynamic_adjacency_array.cpp:312]: as we have discussed ...

**Review by kittobi1992 — CHANGES_REQUESTED** [review]

File: mt-kahypar/datastructures/dynamic_adjacency_array.cpp [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+  Header& head_v = header(v);
+  for (HypernodeID current_v: headers(v)) {
+    const HyperedgeID last = firstInactiveEdge(current_v);
+    for ( HyperedgeID curr_edge = firstActiveEdge(current_v); curr_edge < last; ++curr_edge ) {
+      Edge& e = edge(curr_edge);
+      if (e.isValid() && e.isSinglePin()) {
+        ASSERT(e.source == v);
+        e.setValid(false);
+        --head_v.degree;
+      } else if (e.isValid()) {
```
[the remark this answers is not in the document]
Comment (N-Maas) [reply mt-kahypar/datastructures/dynamic_adjacency_array.cpp:229]: as discussed

[the remark this answers is not in the document]
Comment (N-Maas) [reply mt-kahypar/datastructures/dynamic_adjacency_array.cpp:229]: as discussed

**Pull request merged**